### DISCUSSION TITLE
Compare Drafts / Diff Tool for Entries

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8135] [Ideas]
> Created by: David Ritter (david.ritter@spinthis.com) on 2015/05/18 19:35:25 +0000
> Votes at time of UserVoice import: 14

-----------------------------------

I love Craft's Drafts.  Compared to everything else out there, the live previewing, combined with sharing of unpublished drafts work extremely well.

The only thing missing is when comparing a couple drafts or a draft with the live version, it'd be great to see what the differences are.

One simple option I'd love to see is having a "compare drafts" mode; when invoked, it'd pull up a live preview with forwards and backwards buttons where you can flip between the versions and spot the differences yourself between versions. 

Looking at how other CMS do it, for example, ProcessWire has their upcoming ProDrafts feature where it shows only the fields that where changed (for example, in the case of an asset or relational field).

https://processwire.com/blog/posts/processwire-2.6.13-and-a-preview-of-prodrafts/

Some kind of diff tool for html / text fields would be superb. 

It would really interesting if it also worked with live preview. Use something like John Resig's javascript diff tool on the 2 versions and then show the differences in the live preview. 

http://ejohn.org/projects/javascript-diff-algorithm/

Handling the changed assets and matrix in the live preview would be the most challenging part I would guess, among other things, but I believe P&T could pull it off in a way that made sense.

**Comment by angrybrad** [comment]
> Posted by outline4 (s.a@outline4.ch) on 2015/03/11 07:13:27 +0000
-----------------------------------

Several of my customers were asking for this. Especially when you have some sort of workflow where publishers have to put the entries online.

**Comment by lucytate** [comment]
Has this been added or will this be added soon?

**Comment by brandonkelly** [comment]
@justahutch Craft 3.4 is going to start showing which fields were edited within the draft (see https://craftcms.com/blog/craft-34). We would like to add a full-fledged comparison view down the road as well.

**Comment by binarymonkey84** [comment]
@brandonkelly Will there be audit trail type functionality to go with the diff view? Often we need to view a list of what has changed within a specific time frame, and who changed it. Filtering the list and clicking to view a diff in a quick view modal would be ideal.

**Comment by brandonkelly** [comment]
@binarymonkey84 That’s a possibility down the road.

**Comment by andreimoment** [comment]
Diff view on entry level would be very useful. 

We moved to Craft from an in-house CMS which had it and the editors who work on dozens of revisions for each document are feeling the pain of having to re-read the full content instead of just comparing the differences.

Looking forward to v.4

**Comment by andreimoment** [comment]
To add to the conversation - field-level changes are useful, but not sufficient - especially in the case of page-long text content in a single field. What would be useful for our editorial team is similar to git's diff algorithms - showing the differences on a per character, word, or line basis.

In the example text below you can see how an editor would have to carefully proofread both sides in order to spot the differences. This is tedious, time consuming, and error prone - a perfect use for software.

![CleanShot 2023-07-19 at 13 28 37](https://github.com/craftcms/cms/assets/6190/7c9bd9be-5a3e-4059-abb5-07af6f4972ef)

I am aware of the difficulties of comparing text in the context of html documents which may include formatting. For editorial purposes stripping all formatting would be a viable option.

**Reply by brandonkelly** [reply]
Yeah, we definitely should add this. There are libraries that help with it. The bigger challenge is figuring out how to present it for nested Matrix values, etc.

**Reply by rob-c-baker** [reply]
I did something like this for a project that is currently being driven by Craft. We treat each complex (Matrix etc) field as a collection of smaller simple fields (plain text, number, date, etc). We use a combination of visually highlighting any simple field that is different to the live entry and then a small button next to the field that opens a pop-out  / modal style diff looking not dissimilar to the screen grab above that shows the actual detailed difference between the fields.

In terms of code the diffs are worked out recursively and initially only present themselves as a different highlight on the field (much like a validation error) and therefore don't interfere with the interface much and work on fields nested however deep. Editors can see, at a glance which fields are different and can chose to look at a detailed diff of the fields if they want. We are not quite at the point of merging the contents of fields together (akin to a git merge) because it's out of scope for that project but it would be an interesting addition.

This approach works well for us but I get that it might not work quite as well in a more general way.

**Reply by andreimoment** [reply]
@brandonkelly my first take on this is to simplify as much as possible. 

What if we took the rendered content of both revisions, stripped all tags, equalized the new lines and did a diff? 

As a bonus, what if we json-infield all matrix content to retain the field names and then again stripped all tags for each content block. This would provide the field names and possibly allow linking back for editing.

Text diff could be shown in the preview pane.

**Comment by arnotenhagen** [comment]
@brandonkelly Is this feature request still planned for development? We have multiple clients asking for this. It would be great if this could be included in Craft 6.
